### ISSUE TITLE
Typecheck Failing

### ISSUE THREAD

**Issue by wagenet** [castwide/solargraph#366]
Command: `solagraph typecheck`

Output:

```
Traceback (most recent call last):
        26: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/bin/ruby_executable_hooks:24:in `<main>'
        25: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/bin/ruby_executable_hooks:24:in `eval'
        24: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/bin/solargraph:23:in `<main>'
        23: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/bin/solargraph:23:in `load'
        22: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/bin/solargraph:5:in `<top (required)>'
        21: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/thor-1.0.1/lib/thor/base.rb:485:in `start'
        20: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/thor-1.0.1/lib/thor.rb:392:in `dispatch'
        19: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/thor-1.0.1/lib/thor/invocation.rb:127:in `invoke_command'
        18: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/thor-1.0.1/lib/thor/command.rb:27:in `run'
        17: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/shell.rb:146:in `typecheck'
        16: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/shell.rb:146:in `each'
        15: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/shell.rb:148:in `block in typecheck'
        14: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/type_checker.rb:43:in `problems'
        13: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/type_checker.rb:77:in `method_tag_problems'
        12: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/type_checker.rb:77:in `each'
        11: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/type_checker.rb:78:in `block in method_tag_problems'
        10: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/type_checker.rb:88:in `method_return_type_problems_for'
         9: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/pin/base_method.rb:33:in `typify'
         8: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/pin/base_method.rb:120:in `typify_from_super'
         7: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/api_map.rb:378:in `get_method_stack'
         6: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/api_map.rb:319:in `get_methods'
         5: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/api_map.rb:726:in `resolve_method_aliases'
         4: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/api_map.rb:726:in `each'
         3: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/api_map.rb:727:in `block in resolve_method_aliases'
         2: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/api_map.rb:751:in `resolve_method_alias'
         1: from /Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/api_map.rb:751:in `new'
/Users/peterwagenet/.rvm/gems/ruby-2.7.1/gems/solargraph-0.39.16/lib/solargraph/pin/yard_pin/method.rb:18:in `initialize': undefined method `namespace' for #<Hash:0x00007f89ebbac698> (NoMethodError)
```

The error raises here:

https://github.com/castwide/solargraph/blob/master/lib/solargraph/pin/yard_pin/method.rb#L17

These are the arguments that `initialize` is called with:

```rb
{:code_object=>
  {:location=>
    #<Solargraph::Location /Users/peterwagenet/Development/Tilde/Skylight/direwolf-client/lib/direwolf/ext/active_record.rb, #<Solargraph::Range #<Solargraph::Position 3, 2> to #<Solargraph::Position 3, 53>>>,
   :closure=>
    #<Solargraph::Pin::Namespace at #<Solargraph::Location /Users/peterwagenet/Development/Tilde/Skylight/direwolf-client/lib/direwolf/ext/active_record.rb, #<Solargraph::Range #<Solargraph::Position 2, 0> to #<Solargraph::Position 22, 3>>>>,
   :name=>"structure_dump_without_cleanup",
   :comments=>"",
   :scope=>:instance,
   :visibility=>:public,
   :parameters=>
    [#<Solargraph::Pin::Parameter at nil>,
     #<Solargraph::Pin::Parameter at nil>]},
 :name=>nil,
 :scope=>nil,
 :visibility=>nil,
 :closure=>nil,
 :spec=>nil}
```

**Comment by wagenet** [comment]
It looks like `YardPin::Method` has a different signature than `Pin::Method` and that api_map doesn't account for that.

**Comment by wagenet** [comment]
It looks like this happens because we are monkeypatching a gem. Here's what's in lib/direwolf/ext/active_record.rb:

```rb
class ActiveRecord::Tasks::PostgreSQLDatabaseTasks
  alias structure_dump_without_cleanup structure_dump
  def structure_dump(filename, extra_flags)
    structure_dump_without_cleanup(filename, extra_flags)
    remove_sql_extension_comments(filename)
  end

  private

  def remove_sql_extension_comments(filename)
    tempfile = Tempfile.open("uncommented_structure.sql")
    begin
      File.foreach(filename) do |line|
        tempfile << (line.start_with?("COMMENT ON EXTENSION") ? "-- #{line}" : line)
      end
    ensure
      tempfile.close
    end
    FileUtils.cp(tempfile.path, filename)
  end
end
```

If I ignore this file in .solargraph.yml it works without issue.

**Comment by castwide** [comment]
Thanks for all the details. I fast-tracked a fix and released it in v0.39.17.

**Comment by wagenet** [comment]
Thanks! While investigating this, I also realized that I should convert this to be `prepend`ing a module. This also resolved things for solargraph.

**Comment by opiation** [comment]
> Thanks for all the details. I fast-tracked a fix and released it in v0.39.17.

@wagenet or @castwide, can this be closed now that a fix was released?

**Comment by wagenet** [comment]
@opiation should be fine to close.
